### PR TITLE
Odometry

### PR DESCRIPTION
**Pull request by bart-de-jonge**
Added odometry publisher

### CODE REVIEW

File: src/odometry_publisher.cpp [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    // publish the message
+    odom_pub.publish(odom);
+}
+
+/**
+ * Returns the absolute value of a double.
+ * @param x - the input double
+ * @return - the absolute value of the input
+ */
+double OdometryPublisher::myAbs(double x) {
```
Comment (hungnguyen94) [inline src/odometry_publisher.cpp:90]: What's wrong with `std::abs`?

File: src/dpr4_base.cpp [hunk 2]
```diff
@@ -15,8 +15,14 @@ DPR4Base::DPR4Base()
...
-    wheelBase = 0.537d;
-    wheelDiameter = 0.292d;
+
+    //Settings garbage robot:
+    // wheelBase = 0.537d;
+    // wheelDiameter = 0.292d;
+
+    // Settings host robot:
+    wheelBase = 29.3616564746d;
+    wheelBase = 54.3d;
```
Comment (hungnguyen94) [inline src/dpr4_base.cpp:25]: Je mist `wheelDiameter`. Waarom trouwens in centimeters?

File: src/odometry_publisher.cpp (hunk 1 above)
Comment (bart-de-jonge) [reply src/odometry_publisher.cpp:90]: It's int only

File: src/odometry_publisher.cpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+    // No rotation
+    geometry_msgs::Quaternion cam_quat = tf::createQuaternionMsgFromYaw(0.0);
+
+    // Transform for kinect, x is forwards, y left, z upwards.
+    geometry_msgs::TransformStamped camera_trans;
+    camera_trans.header.stamp = curTime;
+    camera_trans.header.frame_id = "base_link";
+    camera_trans.child_frame_id = "camera_depth_frame";
+    camera_trans.transform.translation.x = 0.08;
```
Comment (Markvanderuit) [inline src/odometry_publisher.cpp:44]: Just a sidenote, but I think I'll extract this back to our launch file tomorrow (given that the issues we had that caused us to put it here are solved), as this transform would never work on the other group's kinect. At least not if you hardcode it.

File: src/odometry_publisher.cpp [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    // Again, no rotation
+    geometry_msgs::Quaternion odom_quat = tf::createQuaternionMsgFromYaw(th);
+
+    // Transform for odom/base_link, x and y inverted because different axis
+    // used in our calculations of odom.
+    geometry_msgs::TransformStamped odom_trans;
+    odom_trans.header.stamp = curTime;
+    odom_trans.header.frame_id = "odom";
+    odom_trans.child_frame_id = "base_link";
+    odom_trans.transform.translation.x = y/100.0;
```
Comment (Markvanderuit) [inline src/odometry_publisher.cpp]: maybe define 100.0 some where as a meter, makes it more readable.

File: src/odometry_publisher.cpp [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+/**
+ * Returns the absolute value of a double.
+ * @param x - the input double
+ * @return - the absolute value of the input
+ */
+double OdometryPublisher::myAbs(double x) {
+    if(x < 0) {
+        return -1*x;
```
Comment (Ezlo123) [inline src/odometry_publisher.cpp]: return myabs(x) saves you 3 lines.

File: src/odometry_publisher.cpp [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    rightPos = newRightPos;
+    double dLeftDistance = dLeftPos * base->getWheelDiameter() / 2.0d;
+    double dRightDistance = dRightPos * base->getWheelDiameter() / 2.0d;
+    std::cout << "dRightDistance: " << dRightDistance << std::endl;
+    leftDistance += dLeftDistance;
+    rightDistance += dRightDistance;
+
+    // Calculate radius
+    double radius;
+    if(myAbs(dLeftDistance - dRightDistance) < 0.00001d) {
```
Comment (Markvanderuit) [inline src/odometry_publisher.cpp:143]: (A) the d in 0.000....001d and other cases is redundant, default floating point values in cpp are double (most of the time)
(b) maybe define the number somewhere.

File: src/odometry_publisher.cpp [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        radius = 0;
+    } else {
+        radius = base->getWheelBase() / 2.0 * ((dLeftDistance + dRightDistance) / (dLeftDistance - dRightDistance));
+        if (radius == 0){
+        	turnInPlace = true;
+        }
+    }
+    if(radius < 0) radius = -1*radius;
```
Comment (Markvanderuit) [inline src/odometry_publisher.cpp]: radius = myabs(radius), IMO that's slightly more readable.

File: src/odometry_publisher.cpp [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        std::cout << "Left distance: " << leftDistance << std::endl;
+        std::cout << "Right distance: " << rightDistance << std::endl;
+        std::cout << "Radius: " << radius << std::endl;
+        std::cout << "dTheta: " << dTheta << std::endl;
+        std::cout << "dx: " << dx << std::endl;
+        std::cout << "x " << x << std::endl;
+        std::cout << "y " << y << std::endl;
+        std::cout << "th " << th << std::endl;
+        std::cout << "th " << 360 * (th / (2 * 3.1415)) << std::endl;
+        std::cout << "\n";
```
Comment (Markvanderuit) [inline src/odometry_publisher.cpp]: Cool logging, but may I suggest
(A) replace every std::endl with '\n'
(B) replace the single final '\n' with cout << '\n' << endl.
TLDR, \n is fast, endl flushes the output buffer. So shit lots of newlines, then when done flush the toilet.

**Review by Markvanderuit — COMMENTED** [review]
Just some minor issues, mostly code cleanup.

File: src/odometry_publisher.cpp (hunk 4 above)
Comment (bart-de-jonge) [reply src/odometry_publisher.cpp]: Is that really necessary? I think most people remember a meter being 100 centimeters, will comment it though

File: src/odometry_publisher.cpp (hunk 6 above)
Comment (bart-de-jonge) [reply src/odometry_publisher.cpp:143]: "most of the time" --> there you go.
